Use temp variable for `this` in virtual method dispatch tables

owen-mc-diffblue opened the pull request:
This is so that value-set filtering in symex will
filter elements which don't have the correct class
identifier out of the value-set for `this`.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Make symbol_table parameter non-const

We will add a symbol to it, so it cannot be const.

Commit:
Use temp variable for `this` in dispatch tables

This is so that value-set filtering in symex will
filter elements which don't have the correct class
identifier out of the value-set for `this`.

Commit:
Only create temp var for this arg if needed

If the this argument doesn't contain a dereference then value-set
filtering will work on it even without creating a temporary
variable

tautschnig commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -144,6 +146,63 @@ static void create_static_function_call(
+  const auto dereference_it = std::find_if(
+    argument_for_this.depth_cbegin(),
+    argument_for_this.depth_cend(),
+    [](const exprt &expr) { return can_cast_expr<dereference_exprt>(expr); });
```
You could use `const bool argument_for_this_contains_a_dereference = has_subexpr(argument_for_this, ID_dereference);` If you don't like the way `has_subexpr` is implemented then you might still use `has_subexpr` here, but also replace the implementation of `has_subexpr` by your code.

smowton commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -144,6 +146,63 @@ static void create_static_function_call(
+    // correct class identifier.
+    symbolt &temp_symbol = get_fresh_aux_symbol(
+      argument_for_this.type(),
+      "remove_virtual_functions", // TODO: pass function name in to use here?
```
Yes, see for example `remove_instanceof.cpp` is doing this (and its callers conveniently (a) live alongside remove-virtuals callers and (b) already have some way to lay their hands on the function ID. The mode can then be retrieved by looking it up in the symbol table.

tautschnig replied:
Yes, please.

smowton reviewed (commented):
LGTM apart from the noted TODO

tautschnig commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -144,6 +146,63 @@ static void create_static_function_call(
+      "remove_virtual_functions", // TODO: pass function name in to use here?
+      "this_argument",
+      vcall_source_loc,
+      ID_java, //TODO: get this from somewhere
```
Require that name of the `goto_program` that's being processed is passed in as well (just like `remove_function_pointers` does this, search for `function_id`). You can then look up the symbol corresponding to `function_id` and use the `mode` of that symbol.

tautschnig commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -144,6 +146,63 @@ static void create_static_function_call(
+    new_code_for_this_argument.add(
+      goto_programt::make_assertion(null_check, vcall_source_loc));
+    new_code_for_this_argument.add(
+      goto_programt::make_assumption(null_check, vcall_source_loc));
```
That's only valid for Java, is it?

owen-mc-diffblue replied:
You're right, this is far too java-specific. It's needed because without it the local safe pointer analysis wasn't realising that `this` definitely wasn't null after the null check. Other options include:
1) making the local safe pointer analysis a teensy bit smarter;
2) checking for a null check in the two instructions above the virtual function call, and only inserting a null check here if we find one.

owen-mc-diffblue replied:
I've done option 2.

tautschnig reviewed (commented):
Just some very general comments, I'll defer to others for approval. My understanding of this bit of code isn't good enough.

Commit:
Use correct function identifier in symbol name

Commit:
Copy directly preceding ASSUMEs

Copy directly preceding ASSUMEs, changing the argument for `this` to
the new temporary variable for `this`. This is mainly aimed at null
checks, to make the local safe pointer analysis work. We stop as
soon as we find an instruction that is not an ASSERT or ASSUME.

Commit:
Test temp var for this argument in virtual function calls

A temporary variable should be created for the `this` argument of a
virtual function call when there is a dispatch table and the `this`
argument involves a dereference.

owen-mc-diffblue commented:
Okay, I have addressed all review comments and fixed the CI issues that that caused. This is ready for re-review.

smowton commented on jbmc/regression/jbmc/virtual_temp_var_for_this_argument/test_one_candidate.desc:
```diff
@@ -0,0 +1,12 @@
+--
+The temporary variable `this_argument` should not be created, because it
+is only created when there is a dispatch table and the `this` argument has a
+dereference in it. Both conditions are false.
```
`. Both conditions are false.` -> `, neither of which is the case here`

smowton commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -144,11 +155,126 @@ static void create_static_function_call(
+      }
+    }
+
+    if(!changed)
```
Probably don't do this -- for example, with
```
ASSUME(x->y->z != null);
ASSUME(some_param != null);
tmp_this = x->y->z;
```
we should keep the assumption concerning `x->y->z`

We might never currently generate any intervening assumptions, but given it's easy to do so we should probably tolerate them if present.

owen-mc-diffblue replied:
Let me just check I understand you correctly: the thing you're saying we shouldn't do is stopping when we encounter an assert or assume which doesn't contain `argument_for_this`; you want us to keep on going until we find an instruction which isn't an assert or an assume (or we hit the beginning of the function).

smowton commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -144,11 +155,126 @@ static void create_static_function_call(
+/// If there are ASSERT or ASSUME instructions involving \p argument_for_this
+/// in the instructions directly preceding the virtual function call then we
+/// duplicate them for \p temp_var_for_this. This is mainly aimed at null
+/// checks, to make the local safe pointer analysis work. We stop as soon as we
```
`to make the local safe pointer analysis work` -> `because \ref local_safe_pointerst would otherwise lose track of known-not-null pointers due to the newly introduced assignment`

smowton commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -144,11 +155,126 @@ static void create_static_function_call(
+  const source_locationt &vcall_source_loc,
+  goto_programt &new_code_for_this_argument)
+{
+  const bool argument_for_this_contains_a_dereference =
```
Given the variable is longer than the expression, just inline this :)

smowton commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -144,11 +155,126 @@ static void create_static_function_call(
+  {
+    // Create a temporary for the `this` argument. This is so that value-set
+    // filtering in symex will reduce the value-set for `this` to those elements
+    // with the correct class identifier.
```
Provide a reference to a function on `goto_symext` for further information

smowton commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -144,11 +155,126 @@ static void create_static_function_call(
+  {
+    instr_it = std::prev(instr_it);
+
+    if(!(instr_it->type == ASSERT || instr_it->type == ASSUME))
```
Agree to skipping over assertions, but suggest only duplicating assumptions (duplicating assertions would be confusing to a user, as any violation would show up twice)

smowton reviewed (approved):
Looks good, some mainly-cosmetic improvements

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 32b0e624efe13c13f3544db22d9de2481239d864).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104940146

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9dbe4fdb9f9f0da1fedbf77da1c16d5ff44e640f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104983138

peterschrammel commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -23,14 +26,17 @@ class remove_virtual_functionst
 
-  bool remove_virtual_functions(goto_programt &goto_program);
+  bool remove_virtual_functions(
+    irep_idt function_identifier,
```
```suggestion
    const irep_idt &function_id,
```

const & (several occurrences)

Also, I think we settled on calling this `function_id` throughout the codebase.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 77075e680bbdee84c3f2528426700c7b8717e4ef).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105077354
